### DISCUSSION TITLE
Fatal error: Uncaught Error: Class "chillerlan\QRCodeExamples\QRImageWithLogo" not found

### DISCUSSION THREAD

**Discussion by dividerbeam** [chillerlan/php-qrcode#75] [Q&A]
Fatal error: Uncaught Error: Class "chillerlan\QRCodeExamples\QRImageWithLogo" not found in /var/www/html/apps/Qrcode/php-qrcode-main/examples/imageWithLogo.php:42 Stack trace: #0 {main} thrown in /var/www/html/apps/Qrcode/php-qrcode-main/examples/imageWithLogo.php on line 42

Hello, whats problem? in other files all work with require_once $root.'/vendor/vendor3/autoload.php';

```php
<?php
/**
 *
 * @filesource   imageWithLogo.php
 * @created      18.11.2020
 * @author       smiley <smiley@chillerlan.net>
 * @copyright    2020 smiley
 * @license      MIT
 */

namespace chillerlan\QRCodeExamples;
$root = $_SERVER["DOCUMENT_ROOT"];
use chillerlan\QRCode\{QRCode, QROptions};

require_once $root.'/vendor/vendor3/autoload.php';

$data = 'https://mydomain.ru';
/**
 * @property int $logoSpaceWidth
 * @property int $logoSpaceHeight
 *
 * @noinspection PhpIllegalPsrClassPathInspection
 */
class LogoOptions extends QROptions{
	// size in QR modules, multiply with QROptions::$scale for pixel size
	protected int $logoSpaceWidth;
	protected int $logoSpaceHeight;
}

$options = new LogoOptions;

$options->version          = 7;
$options->eccLevel         = QRCode::ECC_H;
$options->imageBase64      = false;
$options->logoSpaceWidth   = 13;
$options->logoSpaceHeight  = 13;
$options->scale            = 5;
$options->imageTransparent = false;

//header('Content-type: image/png');

$qrOutputInterface = new QRImageWithLogo($options, (new QRCode($options))->getMatrix($data));

// dump the output, with an additional logo
echo $qrOutputInterface->dump(null, $root.'/apps/Qrcode/php-qrcode-main/examples/octocat.png');
```

**Comment by codemasher** [comment]
The example namespace is only available in autoload-dev, so you might need to install the packages with `composer install --dev` explicitly, or copy that file over to your project, change the namespace and include it from there.

**Reply by dividerbeam** [reply]
composer require chillerlan/php-qrcode --dev   correctly?

**Reply by dividerbeam** [reply]
![image](https://user-images.githubusercontent.com/52495131/126030999-f3b7120c-1221-484c-834f-66f69155be33.png)
i already have QRImageWithLogo.php in folder, but it not work

**Reply by dividerbeam** [reply]
in imageWithLogo.php and QRImageWithLogo.php already has same namespace

**Reply by codemasher** [reply]
The question is in which project folder you're working. If you're working with a clone of this repository directly (which is what it looks like from your screenshot), then you would run `composer install` from its root directory and it should work fine.
In case you want to install it in your own project, you would either run `composer require chillerlan/php-qrcode` or add it to your own project's composer.json under "require" and then run install. However, i cannot tell if your own autoloader in "vendor3" properly includes the files.
Also i was wrong about the `--dev` switch, which would add a package to the dev-dependencies; `compser install` includes dev dependencies by default and you need to add `--no-dev` to prevent it from doing so.

**Reply by dividerbeam** [reply]
please help me, I can't connect the class in any way QRImageWithLogo.php, I didn't understand your answer

**Comment by codemasher** [comment]
You understand that i cannot help you with the little information that i have? Have you tried a clean install/clone of this library and installed the dependencies like described in the readme? Where does the "vendor3" subdirectory come from? From what it seems you have mismatching directories for you dependencies.

**Reply by dividerbeam** [reply]
i have several vendor dirs because i use several service, if they all in one vendor dir, they interfere with each other, can you help to fix that?

and anyway, i was try to delete all other vendor dirs and left only one vendor for your QR codes, it wasn't work too, same errors are issued

and all other files generates QR codes correctly, problems only in QR with logo, only in imageWithLogo.php

look:
![image](https://user-images.githubusercontent.com/52495131/126223770-2771355a-c180-46e3-8e79-4d5c1c0d7b01.png)

and result - https://wantu.ru/apps/Qrcode/php-qrcode-main/examples/html.php

if something is wrong with my paths, then all the files should not have worked

**Comment by codemasher** [comment]
Ok, you definitely have a vendor path mismatch problem and it has absolutely nothing to do with this library. Installing dependencies and then moving the vendor folders just will not work as expected because composer hardcodes the library subdirectories in the autogenerated `autoload_*.php` files in `vendor/composer`. If you move those files, the autoloader simply cannot find the subdirectories.
I'm not sure what you mean by "different services" - i assume you mean different libraries that you would require via a `composer.json` file for your project. If for some reason you need to solve interfering dependecies, you may use the [`vendor-dir` option](https://getcomposer.org/doc/06-config.md#vendor-dir) for composer. Further i'd recommend reading through the composer fundamentals.

**Reply by dividerbeam** [reply]
thanks, I'll look into that.

**Reply by dividerbeam** [reply]
i specified vendor-dir and reinstalled chillerlan/php-qrcode to the desired folder, it didn't help

Fatal error: Uncaught Error: Class "chillerlan\QRCodeExamples\QRImageWithLogo" not found in /var/www/html/apps/Qrcode/php-qrcode-main/examples/imageWithLogo.php:43 Stack trace: #0 {main} thrown in /var/www/html/apps/Qrcode/php-qrcode-main/examples/imageWithLogo.php on line 43

![image](https://user-images.githubusercontent.com/52495131/126294671-bace476d-3260-4647-b0b4-522ef9ba6853.png)

![image](https://user-images.githubusercontent.com/52495131/126294812-b70221b9-758b-4694-908c-cf506712d2ab.png)

https://wantu.ru/apps/Qrcode/php-qrcode-main/examples/imageWithLogo.php

**Comment by codemasher** [comment]
See https://github.com/chillerlan/php-qrcode/discussions/78#discussioncomment-1025393

**Comment by codemasher** [comment]
Also, you're installing v2.x - the class you're looking for is only available since 4.3 (with a backport to 3.2).

**Reply by dividerbeam** [reply]
how do I install the desired version?

**Comment by codemasher** [comment]
change this:
![image](https://user-images.githubusercontent.com/592497/126334770-a59969d4-6cf6-4f08-b38f-4b6b5da2dc20.png)

**Reply by dividerbeam** [reply]
I didn't ask what I wanted)) the question is rather why is version 2 there by default? I didn't specify it anywhere, why isn't the latest version installed by itself?

**Reply by dividerbeam** [reply]
![image](https://user-images.githubusercontent.com/52495131/126352054-e5f19138-7e19-42f8-b532-8fa0bbd63033.png)

it chooses itself:  "Using version ^2.0 for chillerlan/php-qrcode"

command 
composer require chillerlan/php-qrcode:4.3
says:
[InvalidArgumentException]
  Package chillerlan/php-qrcode at version 4.3 has a PHP requirement incompatible with your PHP version, PHP extensions and Composer version

**Comment by codemasher** [comment]
~~I have literally no idea. Maybe it uses the version according to your PHP version? I get the latest version when installing vial `composer require` on PHP 7.4.~~

Edit: according to your updated comment, your PHP version does not support the latest and not even v3, which requires PHP 7.2. So it's recommended to upgrade your PHP first as you're running an [EOL PHP](https://www.php.net/supported-versions.php).

**Reply by dividerbeam** [reply]
composer cant see my php version in docker container, now i try install composer inside container

**Comment by codemasher** [comment]
Once again, i'm not PHP support. For further questions about general things that clearly have nocthing to do with this library, check [StackOverflow](https://stackoverflow.com/questions/tagged/php) or the [PHPhelp subreddit](https://www.reddit.com/r/PHPhelp/). Thanks and good luck.

**Reply by dividerbeam** [reply]
QRImageWithLogo not work anyway

Fatal error: Uncaught Error: Call to undefined method chillerlan\QRCode\Data\QRMatrix::setLogoSpace() in /var/www/html/vendor/vendor3/chillerlan/php-qrcode/src/QRImageWithLogo.php:43 Stack trace: #0 /var/www/html/apps/Qrcode/php-qrcode-main/examples/imageWithLogo.php(47): chillerlan\QRCode\QRImageWithLogo->dump(NULL, '/var/www/html/a...') #1 {main} thrown in /var/www/html/vendor/vendor3/chillerlan/php-qrcode/src/QRImageWithLogo.php on line 43

**Comment by dividerbeam** [comment]
Fatal error: Uncaught Error: Class "chillerlan\Settings\SettingsContainerAbstract" not found in /var/www/html/vendor/vendor4/chillerlan/php-qrcode/src/QROptions.php:52

where i can find this class SettingsContainerAbstract?

**Comment by codemasher** [comment]
This library [depends on it](https://github.com/chillerlan/php-qrcode/blob/be3beb936c21fe53a4e7e8f7f3582e9f02443666/composer.json#L29) and it should have been installed automatically ([chillerlan/php-settings-container](https://github.com/chillerlan/php-settings-container)) - you clearly have messed up dependencies. 
The method that produced the error in your previous post is in [v3.2.x](https://github.com/chillerlan/php-qrcode/blob/v3.2.x/src/Data/QRMatrix.php#L585) and [v4.3.x](https://github.com/chillerlan/php-qrcode/blob/be3beb936c21fe53a4e7e8f7f3582e9f02443666/src/Data/QRMatrix.php#L591)

**Reply by dividerbeam** [reply]
![image](https://user-images.githubusercontent.com/52495131/126536461-e0ddb4db-c3d2-4f95-a138-28048d962b47.png)

whats wrong now? all dependencies was instaled
